contoso_manufacturing/developer/webapp-decode/yolov8.py

```diff
@@ -145,39 +154,84 @@ def draw_detections(self, img, box, score, class_id):
+                       f"Input: {self.img_width}x{self.img_height}",
+                       f"Inferencing: {self.input_width}x{self.input_height}",
+                       f"Model: {self.model_name}"]
+        
+        # Define the font style and size
+        font_scale = 1.0
+        font_face = cv2.FONT_HERSHEY_SIMPLEX
+        font_thickness = 1
+        background_color = (0,0,255)    # red
+        font_color = (0, 0, 0)          # black
+        pixel_border = 10
+
+        # Define the starting position for the text (30 pixels from the top left corner)
+        (label_x, label_y) = (30, 30)
+
+        # Loop through each line of text in the label array and draw it on the image
+        for label in label_array:
+            # Calculate the dimensions of the label text
+            (label_width, label_height), _ = cv2.getTextSize(label, font_face, font_scale, font_thickness)
+            self.log(f"Label: {label}, Width: {label_width}, Height: {label_height}")
+
+            # Draw a filled rectangle as the background for the label text, including a border on all sides
+            # Draw the label text on the image
+            # Update the starting position for the next line, including a pixel_border pixel margin between lines
+            cv2.rectangle(img, (label_x - pixel_border, label_y - pixel_border), (label_x + label_width + pixel_border, label_y + label_height + pixel_border), background_color, cv2.FILLED)
+            cv2.putText(img, label, (label_x, label_y + label_height), font_face, font_scale, font_color, font_thickness, cv2.LINE_AA)
+            label_y += (label_height + 2 * pixel_border)
+
     def run(self):
-        if(self.verbose):
-            print("Running detection...")
+        self.log("Running detection...")
 
         self.frame_number += 1
         # If mod = 0, i will get the frame and skip it
-        if self.frame_number % self.skip_rate == 0:
+        if ((self.skip_rate > 0) and (self.frame_number % self.skip_rate == 0)):
             self.cap.read()
             return None
         
+        # Pre-process the image
+        # Perform inference on the preprocessed image; capture the start and end times
+        # Post-processing including drawing bounding boxes
         image_data = self.preprocess()
-
+        time1 = time.time()
         outputs = self.grpc_client.predict({"images": image_data}, self.model_name)
+        time2 = time.time()
         frame = self.postprocess(self.cap.read()[1], outputs)
 
+        # Update metrics used for FPS calculations
+        self.total_inference_time += (time2 - time1)
+        self.total_frames += 1
+
+        # Calculate FPS for both the inferencing step and the final feed  
+        self.inference_fps = self.total_frames / self.total_inference_time
+        self.total_fps = self.total_frames / (time.time() - self.start_time)    # This includes e.g. JPEG encoding in the parent method outside of self.run()
```

Comment:
Calculating the total time in inference function here does not take into account the post processing time for the frame. If we also need to include how much time it takes to draw the bounding boxes/output the frame then we might need to do this after we have drawn the boxes and just before displaying the FPS.